enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,940 @@
+
+Design aims to implement following principles:
+- Keep it simple, optimize later
+- MicroShift does not own the OS or host
+- MicroShift and all its components are versioned, upgraded, and rolled back together
+- Be defensive, fail fast
+- Rely on outside intervention as a last resort
+
+### Non-Goals
+
+* Building MicroShift upgrade graph
+* 3rd party applications' health checks and its data backup or rollback
+  (although end user documentation should be provided)
+* Defining procedures for backup and restore, and upgrading MicroShift on
+  non-ostree systems is left to a future enhancement
+
+## Proposal
+
+### Integration with greenboot
+
+greenboot integrates with systemd, ostree, and grub to provide auto-healing
+capabilities of newly staged and booted deployment in form of reboot:
+if system is still unhealthy after specified amount of reboots, it
+will be rolled back to previous ostree deployment (commit).
+
+Because greenboot already exists as an integral part of Red Hat Device Edge
+systems, we will integrate with it, rather than creating a new system.
+
+greenboot determines system health with health check scripts and MicroShift
+already provides such script. For more information about greenboot and current
+MicroShift integration see [Integrating MicroShift with Greenboot](./microshift-greenboot.md).
+
+After health check, either "green" (system is healthy) or "red" (system is unhealthy)
+scripts are executed. MicroShift will provide "green" and "red" scripts which
+will persist the overall system's health for the current ostree commit.
+
+### Triggers for greenboot failures
+
+System images can introduce different types of changes, including:
+1. New OS content unrelated to MicroShift
+2. Different configuration settings for MicroShift
+3. Different versions of MicroShift (higher or lower)
+4. Different versions of applications running on MicroShift (higher or lower)
+5. More, fewer, or different applications
+
+Any of those transitions could result in a greenboot failure.
+
+Because MicroShift cannot detect the cause of the failure, and cannot influence
+how greenboot handles the failure, all failures will be handled by reverting
+to a previous known-good state for MicroShift's data.
+
+### Version change support
+
+- Because we want to maintain upgrade expectations with Kubernetes and OpenShift,
+  we will only support changing versions one Y version at a time (x.y to x.y+1).
+- Because we cannot guarantee that the data formats between Y versions are
+  compatible after an upgrade, we will only support rolling back to a previous
+  Y version when restoring from a backup.
+- Because we may need to support "manual" changes to correct for regressions
+  within a Y version, and because we expect the storage format and other data
+  to be forward and backward compatible within a Y version, we will support
+  changing from any Z version to any other Z version for the same version of Y,
+  including downgrading.
```

Comment:
> support it without requiring a backup?

Yes, backup would be needed to "downgrade" Y stream (in form of rollback that would restore the backup).
Idea is to "support" (maybe allow or "not prohibit" is better term) scenarios like: run microshift 4.14.5 and then stage, boot, and run image with ushift 4.14.1.
Since, like you mentioned in other comment, we cannot retroactively amend released software (and its "block list") , if there's a problem when "Z-downgrading" from 4.14.5 to 4.14.1 but it works fine to 4.14.2, then it can be only documented when released 4.14.5